### PR TITLE
added beta 7 coreclr dockerfile

### PR DESCRIPTION
**Pull request by jchannon**
This PR provides a beta 7 coreclr dockerfile

This addresses #83

### CODE REVIEW

**Comment by akoeplinger** [comment]
You'll also need these packages (per [docs](http://docs.asp.net/en/latest/getting-started/installing-on-linux.html#install-the-net-execution-environment-dnx)):

`libunwind8, gettext, libssl-dev, libcurl3-dev, zlib1g`

**Comment by jchannon** [comment]
@akoeplinger cheers, added them in, `docker build` worked locally however before adding them in

**Comment by akoeplinger** [comment]
@jchannon sure, cause that doesn't try to run any apps ;) dnu restore needs libcurl e.g.

**Comment by jchannon** [comment]
gotcha, damn it failed build

**Comment by akoeplinger** [comment]
@jchannon 

```
E: Unable to locate package libunwind8
```

That package is only available starting from Debian Jessie, you should switch to that from wheezy.

**Comment by jchannon** [comment]
Failed again

**Comment by akoeplinger** [comment]
> [CI] Sample 'HelloWeb' not found for Docker image '1.0.0-beta7-coreclr' .ci/run-with-all-tags.sh . HelloWeb returned exit code 1

I think the CI script needs to be modified to strip off the `-coreclr` at the end [here](https://github.com/aspnet/aspnet-docker/blob/45f3251ada39c7525493f9b6f45c9b77f1d8904c/.ci/build-app-image.sh#L18).

**Comment by jchannon** [comment]
Is it taking it from my new folder name I wonder

**Comment by akoeplinger** [comment]
Yes it does, it gets all tags (== folders) [here](https://github.com/aspnet/aspnet-docker/blob/45f3251ada39c7525493f9b6f45c9b77f1d8904c/.ci/find-tags.sh#L8) and passes it down [here](https://github.com/aspnet/aspnet-docker/blob/45f3251ada39c7525493f9b6f45c9b77f1d8904c/.ci/run-with-all-tags.sh#L12)

**Comment by jchannon** [comment]
I've stripped it but i'm not sure thats correct as I think we'll need a tag for coreclr. Happy to undo last commit and then take advice on path forward from MS team

**Comment by akoeplinger** [comment]
I think you just need to just strip it at [build-app-image.sh#L18](https://github.com/aspnet/aspnet-docker/blob/45f3251ada39c7525493f9b6f45c9b77f1d8904c/.ci/build-app-image.sh#L18) otherwise it'd take the same `1.0.0-beta7` image twice.

**Comment by jchannon** [comment]
Same in build-all-tags too though

**Comment by akoeplinger** [comment]
No, in that case we want it to be `1.0.0-beta7`. We don't want it in build-app-image.sh as there's no folder called -coreclr in [aspnet/home samples](https://github.com/aspnet/Home/tree/dev/samples).

**Comment by jchannon** [comment]
That passed but I think when it validated the beta7-coreclr helloweb/mvc it then used the beta7 dockerfile. I'll leave it up to MS now to pull this branch and make the necessary changes or point our what needs to change and I'll happily do it

File: .ci/build-app-image.sh [hunk 1]
```diff
@@ -12,7 +12,7 @@ BASE_IMAGE=$1
 TEST_APP=$2
 TAG=$3
 VERSION=$4
-
+VERSION=$(echo $VERSION | sed -e "s/-coreclr//")
```
Comment (ahmetalpbalkan) [inline .ci/build-app-image.sh]: you might wanna do this `s/-coreclr$//` :)

**Comment by ahmetalpbalkan** [comment]
@jchannon yeah looks like we need to refactor CI system based on that. 

Can you please also update `1.0.0-beta7/README.md` to mention this image?

File: .ci/build-app-image.sh (hunk 1 above)
Comment (akoeplinger) [inline .ci/build-app-image.sh]: or even `VERSION=${VERSION%-coreclr}` :)

Comment (ahmetalpbalkan) [inline .ci/build-app-image.sh]: actually let's extract `$SAMPLES_REPO/samples/$VERSION/$TEST_APP` into a variable (it repeats at least twice) and handle it there inline like `".../${VERSION%-coreclr}/..."`

**Comment by superlogical** [comment]
I'll go one even better!
http://basicweb-1835675763.us-east-1.elb.amazonaws.com:5000/

Deployed a test app using that Docker image using a tool called http://www.convox.com/

@jchannon

**Comment by jchannon** [comment]
@ahmetalpbalkan do you want me to add a coreclr sample [here](https://github.com/aspnet/Home/tree/dev/samples) so the tests in this repo pick up the correct HelloWorld samples?

**Comment by jchannon** [comment]
@superlogical i see from that you hardcoded the path to dnu/dnx? I've tried building HelloWeb/Mvc locally and it says dnu/dnx isnt in the PATH when it clearly is

File: 1.0.0-beta7-coreclr/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,25 @@
+FROM debian:jessie
```
Comment (davidfowl) [inline 1.0.0-beta7-coreclr/Dockerfile]: ?

Comment (jchannon) [inline 1.0.0-beta7-coreclr/Dockerfile]: a later debian release than wheezy with the dependencies we need in it

File: 1.0.0-beta7-coreclr/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+ENV DNX_VERSION 1.0.0-beta7
+ENV DNX_USER_HOME /opt/dnx
+
+RUN apt-get -qq update && apt-get -qqy install unzip curl libunwind8 gettext libssl-dev libcurl3-dev zlib1g
+
+RUN curl -sSL https://raw.githubusercontent.com/aspnet/Home/dev/dnvminstall.sh | DNX_USER_HOME=$DNX_USER_HOME DNX_BRANCH=v$DNX_VERSION sh
+RUN bash -c "source $DNX_USER_HOME/dnvm/dnvm.sh \
+	&& dnvm upgrade -r coreclr"
```
Comment (vlesierse) [inline 1.0.0-beta7-coreclr/Dockerfile]: I think it is better do it like the other versions. This implicitly downloads the latest version of the runtime and doesn't set the alias which is required for the PATH.

```bash
RUN bash -c "source $DNX_USER_HOME/dnvm/dnvm.sh \
    && dnvm install $DNX_VERSION -r coreclr -a default \
    && dnvm alias default | xargs -i ln -s $DNX_USER_HOME/runtimes/{}$DNX_USER_HOME/runtimes/default"
```

**Comment by vlesierse** [comment]
I have to agree with @ahmetalpbalkan that tagging the image with `coreclr-1.0.0-beta7` is a better. Shouldn't reflect this in the folder structure as well?

**Comment by jchannon** [comment]
Helloworld samples work with last commit

**Comment by superlogical** [comment]
@jchannon Yeah I hard coded the path to dnu and dnx to get it working. I wish dnu/dnx and dnvm was a single binary

**Comment by akoeplinger** [comment]
@superlogical FYI dnu and dnx are already a single binary (there's dnu.cmd that just calls into dnx.exe with some arguments). dnvm needs to be separate and unmanaged code (it's a powershell/bash script), because where would the dnx that you need to download a new dnx come from? :smile:

File: 1.0.0-beta7-coreclr/Dockerfile (hunk 2 above)
Comment (ahmetalpbalkan) [inline 1.0.0-beta7-coreclr/Dockerfile]: @jchannon can you please check if we still need to compile libuv ourselves in jessie? Historically in #22 we started to compile this ourselves, if the libuv in jessie repos is good enough then we can move all images to jessie in a separate PR.

cc: @akoeplinger

**Comment by ahmetalpbalkan** [comment]
@jchannon this ready to go?

**Comment by jchannon** [comment]
yup, go go go but i imagine youll have to merge the sample first

**Comment by ahmetalpbalkan** [comment]
This made me wonder how does this build pass if the sample is not merged.

**Comment by jchannon** [comment]
because it uses the 1.0.0-beta7 version. thats what i saw in the build log anyway

**Comment by jchannon** [comment]
```
[CI] Verifying 'HelloWeb' app with 'aspnet-ci:1.0.0-beta7-coreclr'
[CI] Injecting Dockerfile to project HelloWeb...
total 16
drwxrwxr-x 1 ubuntu ubuntu  126 Sep  4 13:24 .
drwxrwxr-x 1 ubuntu ubuntu   98 Sep  4 13:24 ..
-rw-rw-r-- 1 ubuntu ubuntu  172 Sep  4 13:24 Dockerfile
-rw-rw-r-- 1 ubuntu ubuntu 1144 Sep  4 13:24 HelloWeb.xproj
-rw-rw-r-- 1 ubuntu ubuntu  985 Sep  4 13:24 project.json
drwxrwxr-x 1 ubuntu ubuntu   38 Sep  4 13:24 Properties
-rw-rw-r-- 1 ubuntu ubuntu  340 Sep  4 13:24 Startup.cs
drwxrwxr-x 1 ubuntu ubuntu   18 Sep  4 13:24 wwwroot
Using existing Dockerfile in the sample.
Dockerfile:
FROM microsoft/aspnet:1.0.0-beta7

COPY project.json /app/
WORKDIR /app
RUN ["dnu", "restore"]
COPY . /app

EXPOSE 5004
ENTRYPOINT ["dnx", "-p", "project.json", "kestrel"]
```

**Comment by jchannon** [comment]
I think its because of the `build-app-image.sh` as remove the -coreclr suffix so it then uses the 1.0.0-beta7 instead of 1.0.0-beta7-coreclr

File: 1.0.0-beta7-coreclr/Dockerfile (hunk 2 above)
Comment (akoeplinger) [inline 1.0.0-beta7-coreclr/Dockerfile]: @ahmetalpbalkan yeah, jessie is still on 0.10 while we need 1.x: https://packages.debian.org/jessie/libuv-dev

**Comment by ahmetalpbalkan** [comment]
@jchannon for now my suggestion is, just like we trim the -coreclr from the dirname in the CI scripts, we can "sed --in-place" to replace the image name for coreclr samples.

Or a better one: For coreclr images, we can clone the repo and copy the samples into a -coreclr folder just like you did here and do the replacement there in the CI scripts. It would look just the same to the script then. 

What do you think?

**Comment by jchannon** [comment]
I'm easy, I guess whatever is easiest long term for you to maintain :smile:

**Comment by ahmetalpbalkan** [comment]
Let's go with the second option if it sounds good to you as well. Thank you so much. @jchannon

**Comment by jchannon** [comment]
Just to clarify, the cloning already happens as part of the CI build so you want to make a copy of that directory or clone twice one into a samples directory and one into a sample-coreclr directory? I'm wondering if this is a change you guys might have to do?

**Comment by ahmetalpbalkan** [comment]
@jchannon I think if the tag ends with `-coreclr` you will need to 

    cp -rf samples/${VERSION%-coreclr}  samples/${VERSION}

so we'll have `samples/1.0.0-beta7-coreclr` directory just like the PR you sent to aspnet/Home repo. And then we'd go in there and:

    sed -i '/^FROM/ s/$/-coreclr/'

to append `-coreclr` to the line that begins with `FROM` and replace dockerfile like that.

**Comment by jchannon** [comment]
Actually wouldn't it just be easier of we remove the ${VERSION%-coreclr} so
when its cloned from the samples it will use the 1.0.0-beta7-coreclr
directory and image?

**Comment by ahmetalpbalkan** [comment]
@jchannon my argument is, for now let's not duplicate samples in the aspnet/Home repo. Even with your PR we will need re-tagging the version to include your commit. I don't think the `-coreclr` directory should not exist in `aspnet/Home` for now.

**Comment by ahmetalpbalkan** [comment]
BTW if there's anything blocked on this, we can merge and release the image but I prefer waiting on the fix for the CI script.

File: 1.0.0-beta7/README.md [hunk 4]
```diff
@@ -14,6 +14,7 @@ This project is part of ASP.NET 5. You can find samples, documentation, and gett
 ## Supported tags
 
 * [`1.0.0-beta7`, `latest`  _(1.0.0-beta7/Dockerfile)_](https://github.com/aspnet/aspnet-docker/blob/master/1.0.0-beta7/Dockerfile)
+* [`1.0.0-beta7`, `coreclr`  _(1.0.0-beta7/Dockerfile)_](https://github.com/aspnet/aspnet-docker/blob/master/1.0.0-beta7-coreclr/Dockerfile)
```
Comment (ahmetalpbalkan) [inline 1.0.0-beta7/README.md]: @jchannon this line should just be `coreclr-1.0.0-beta7`, I don't think we need to maintain a `coreclr` tag for now.

**Comment by jchannon** [comment]
I think its best you finish this off

**Comment by ahmetalpbalkan** [comment]
OK, merging. Thanks @jchannon
